sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+
+          @Override
+          public void messageArrived(String topic, MqttMessage message) throws Exception {
+            LOGGER.info("Message arrived");
+            currentTimestamp = Instant.now();
+            queue.put(KV.of(message, currentTimestamp));
+          }
+
+          @Override
+          public void deliveryComplete(IMqttDeliveryToken token) {
+            // nothing to do
+          }
+        });
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      LOGGER.info("Taking from the pending queue ({})", queue.size());
+      try {
+        KV<MqttMessage, Instant> message = queue.take();
+        current = message.getKey().getPayload();
+        watermark = message.getValue();
+        if (current == null) {
+          return false;
+        } else {
+          return true;
+        }
+      } catch (InterruptedException e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      LOGGER.info("Closing MQTT reader");
+      try {
+        if (client != null) {
+          try {
+            client.disconnect();
+          } finally {
+            client.close();
+          }
+        }
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return watermark;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return new UnboundedSource.CheckpointMark() {
+        @Override
+        public void finalizeCheckpoint() throws IOException {
+          // nothing to do
+        }
+      };
+    }
+
+    @Override
+    public byte[] getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write and send a message to a MQTT server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract MqttConnectionConfiguration mqttConnectionConfiguration();
+    @Nullable abstract int qos();
+    @Nullable abstract boolean retained();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMqttConnectionConfiguration(MqttConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    public Write withMqttConnectionConfiguration(MqttConnectionConfiguration configuration) {
```

Comment:
Document all the withBlah() methods on the read and write transforms, documenting any interesting caveats.